lib/rubocop/cop/rspec/metadata_style.rb

```diff
@@ -0,0 +1,197 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Use consistent metadata style.
+      #
+      # @safety
+      #   This cop's autocorrection is unsafe in the case of
+      #   `EnforcedStyle: hash` because it cannot determine whether the trailing
+      #   non-literal argument is a Hash or not.
```

Comment:
What is the case exactly?

Reply:
If you try to autocorrect code like this:

```ruby
hash = { config: true }
describe 'Something', :a, hash
```

you will probably get the following result:

```ruby
hash = { config: true }
describe 'Something', hash, a: true
```

---

That's what I thought at first, but now that I've written the test, I see that `RSpec::Cop::RSpec::Metadata#on_block` is not working as intended 🤔 

```ruby
context 'with symbol metadata with another existing non-literal metadata' do
  it 'registers offense' do
    expect_offense(<<~RUBY)
      hash = { config: true }
      describe 'Something', :a, hash
                            ^^ Use hash style for metadata.
    RUBY

    expect_correction(<<~RUBY)
      hash = { config: true }
      describe 'Something', hash, a: true
    RUBY
  end
end
```

```
$ bundle exec rspec spec/rubocop/cop/rspec/metadata_style_spec.rb:213
Run options: include {:locations=>{"./spec/rubocop/cop/rspec/metadata_style_spec.rb"=>[213]}}

Randomized with seed 34162

RuboCop::Cop::RSpec::MetadataStyle
  with `EnforcedStyle: hash`
    with symbol metadata with another existing non-literal metadata
      registers offense (FAILED - 1)

Failures:

  1) RuboCop::Cop::RSpec::MetadataStyle with `EnforcedStyle: hash` with symbol metadata with another existing non-literal metadata registers offense
     Failure/Error: super
     
       Diff:
       @@ -1,4 +1,3 @@
        hash = { config: true }
        describe 'Something', :a, hash
       -                      ^^ Use hash style for metadata.
       
     # ./spec/support/expect_offense.rb:17:in `expect_offense'
     # ./spec/rubocop/cop/rspec/metadata_style_spec.rb:215:in `block (4 levels) in <top (required)>'

Finished in 0.02466 seconds (files took 0.60568 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/rubocop/cop/rspec/metadata_style_spec.rb:214 # RuboCop::Cop::RSpec::MetadataStyle with `EnforcedStyle: hash` with symbol metadata with another existing non-literal metadata registers offense

Randomized with seed 34162
```

We might want to improve this matcher:

https://github.com/rubocop/rubocop-rspec/blob/2aea1be327407a656766432f467beb9dbaecbff6/lib/rubocop/cop/rspec/mixin/metadata.rb#L26-L28

Reply:
Indeed, `Metadata` only works with literal metadata.
Maybe we should keep it this way?
Then, we can make this cop autocorrect safe.
I'd add a test that this case with a var is ignored.

Reply:
The reasons why I think this matcher can be improved are:

- We can consider any non-trailing metadata as Symbol metadata even if it is not literal
- Not really relevant to this issue, but there is no need to support `str` node here

How about marking the autocorrection as safe by saying that it is not supported if the tailing argument is neither a Hash nor Symbol literal?

That way, users can at least know that there is an offense in that case like this:

```ruby
describe 'Something', :a, b do
                      ^^ Use hash style for metadata.
end
```

I know that this kind of case is quite rare, but one might argue that the use of Symbol metadata should be allowed in this case. Personally, I prefer to consistently assume that this is an offense as it is simpler and easier to understand.